src/lookup.h

```diff
@@ -5,6 +5,22 @@
 
 using namespace std;
 
-SEXP find_promise_in_environment(const SEXP symbol, const SEXP rho2);
+enum class lookup_status {
+    SUCCESS,
+    SUCCESS_ACTIVE_BINDING,
+    FAIL_USER_DEFINED_DATABASE,
+    FAIL_ENVIRONMENT_IS_NIL,
+    FAIL_ARGUMENT_IS_NOT_AN_ENVIRONMENT,
+    FAIL_GLOBAL_CACHE
+};
+
+std::string lookup_status_to_string(lookup_status);
+
+struct lookup_result {
```

Comment:
Why do you prefer returning a `struct` instead of logging `FAIL_` conditions at their place of occurrence and returning the looked up value in case of `SUCCESS_` ? Eventually, these enum values are used to log warnings anyways, isn't it so?

Reply:
This happens because the function is used in various contexts, and I wanted to make it general enough so that it can be re-used without modifications if possible. Returning a struct allows me to act differently when there's an active binding, and possibly, in the future, we can handle the errors better, or mitigate them. It's the sort of mechanism you see with, say, file I/O in C: functions return a value and set an error value somewhere, with the exception that here I don't set a global variable, but stuff it all in a struct.